testsuite/tests/typing-modular-explicits/general.ml

```diff
@@ -0,0 +1,1272 @@
+  <fun>
+Line 5, characters 25-37:
+5 | let typing_order1 f = (f (module Int) 3, id' f)
+                             ^^^^^^^^^^^^
+Error: The signature for this packaged module couldn't be inferred.
+|}]
+
+let typing_order2 f = (id' f, f (module Int) 3)
+
+[%%expect{|
+val typing_order2 :
+  ((module T : Typ) -> T.t -> T.t) ->
+  ((module T : Typ) -> T.t -> T.t) * Int.t = <fun>
+|}, Principal{|
+Line 1, characters 30-31:
+1 | let typing_order2 f = (id' f, f (module Int) 3)
+                                  ^
+Warning 18 [not-principal]: applying a dependent function is not principal.
+
+val typing_order2 :
+  ((module T : Typ) -> T.t -> T.t) ->
+  ((module T : Typ) -> T.t -> T.t) * Int.t = <fun>
+|}]
+
+(** The following test check that tests at module unpacking still happen with
+    modular explicits *)
+
+(* we test that free type variables cannot occur *)
+module type T = sig type t val v : t end;;
+let foo (module X : T with type t = 'a) = X.v X.v;;
+
+[%%expect{|
+module type T = sig type t val v : t end
+Line 6, characters 8-39:
+6 | let foo (module X : T with type t = 'a) = X.v X.v;;
+            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The type of this packed module contains variables:
+       "(module T with type t = 'a)"
+|}]
+
+(* Test principality warning of type *)
+let principality_warning2 f =
+  let _ : ((module T : Typ) -> T.t -> T.t) -> unit = f in
+  f (fun (module T) x -> x)
+
+[%%expect{|
+val principality_warning2 :
+  (((module T : Typ) -> T.t -> T.t) -> unit) -> unit = <fun>
+|}, Principal{|
+Line 3, characters 9-19:
+3 |   f (fun (module T) x -> x)
+             ^^^^^^^^^^
+Warning 18 [not-principal]: this module unpacking is not principal.
+
+val principality_warning2 :
+  (((module T : Typ) -> T.t -> T.t) -> unit) -> unit = <fun>
+|}]
+
+(*
+    Ensure that application of a module-dependent function does not erase the
+    principality of the return type.
+*)
+let should_be_principal (f : (module M : Typ) -> (module N : Typ) -> M.t * N.t)
+  = f (module Int) (module Int)
+
+[%%expect{|
+val should_be_principal :
+  ((module M : Typ) -> (module N : Typ) -> M.t * N.t) -> Int.t * Int.t =
+  <fun>
+|}]
+
+(* This test check that as `t` is private we cannot inline its definition *)
+module type S = sig
+  type t = private int
+  val f : t
+end
+
+let check_escape : _ -> _ = fun (module M : S) -> M.f
+
+[%%expect{|
+module type S = sig type t = private int val f : t end
+Line 6, characters 50-53:
+6 | let check_escape : _ -> _ = fun (module M : S) -> M.f
+                                                      ^^^
+Error: The value "M.f" has type "M.t" but an expression was expected of type "'a"
+       The type constructor "M.t" would escape its scope
+|}]
+
+(* The following test should give a warning only once.
+  Here we test that the structure is typed only once.
+  To achieve this we wrote a code that raises a warning when typed.
+  If the warning is raised twice then that means typing happened twice.
```

Comment:
This testing strategy doesn't really work since there are often options to disable warnings.